central/scannerdefinitions/handler/handler.go

```diff
@@ -350,311 +376,333 @@ func (h *httpHandler) openOfflineBlob(ctx context.Context, blobName string) (*vu
+			continue
 		}
-		defer cleanUp()
-		offlineFile = &vulDefFile{File: vulns, modTime: openedFile.modTime}
-	default:
-		return nil, fmt.Errorf("unknown updater type: %s", t)
+		if strings.HasPrefix(zipF.Name, scannerV4DefsPrefix) {
+			if err := h.handleScannerDefsFile(ctx, zipF, offlineScannerV4DefsBlobName); err != nil {
+				return errors.Wrap(err, "couldn't handle scanner-v4-defs sub file")
+			}
+			log.Debugf("Successfully processed file: %s", zipF.Name)
+			count++
+			continue
+		}
+		// Ignore any other files which may be in the ZIP.
+	}
+	// Just do a simple check for at least one valid file.
+	// Anything stricter may come with incompatibilities.
+	if count > 0 {
+		return nil
+	}
+	return errors.New("scanner defs file(s) not found in upload zip: incorrect zip?")
+}
+
+func (h *httpHandler) handleScannerDefsFile(ctx context.Context, zipF *zip.File, blobName string) error {
+	r, err := zipF.Open()
+	if err != nil {
+		return errors.Wrap(err, "opening compressed file")
 	}
+	defer utils.IgnoreError(r.Close)
 
-	return offlineFile, nil
+	b := &storage.Blob{
+		Name:         blobName,
+		LastUpdated:  protocompat.TimestampNow(),
+		ModifiedTime: protocompat.TimestampNow(),
+		Length:       zipF.FileInfo().Size(),
+	}
+
+	if err := h.blobStore.Upsert(sac.WithAllAccess(ctx), b, r); err != nil {
+		return errors.Wrap(err, "writing scanner definitions")
+	}
+
+	return nil
 }
 
-// openFromArchive returns a file object for a name within the definitions
-// bundle. The file object has a file descriptor allocated on the filesystem, but
-// its name is removed. Meaning once the file object is closed, the data will be
+// openFromArchive returns the associated file for the given name within the ZIP archiveFile.
+// The returned file struct has a file descriptor allocated on the filesystem outside the ZIP, but
+// its name is removed. Meaning: as soon as the file struct is closed, the data will be
 // freed in filesystem by the OS.
-func openFromArchive(archiveFile string, fileName string) (*os.File, func(), error) {
+func (h *httpHandler) openFromArchive(archiveFile string, fileName string) (*os.File, error) {
 	// Open zip archive and extract the fileName.
 	zipReader, err := zip.OpenReader(archiveFile)
 	if err != nil {
-		return nil, nil, errors.Wrap(err, "opening zip archive")
+		return nil, errors.Wrap(err, "opening zip archive")
 	}
 	defer utils.IgnoreError(zipReader.Close)
-	fileReader, err := zipReader.Open(fileName)
+	zipFile, err := zipReader.Open(fileName)
 	if err != nil {
-		return nil, nil, errors.Wrap(err, "extracting")
+		return nil, errors.Wrap(err, "extracting")
 	}
-	defer utils.IgnoreError(fileReader.Close)
+	defer utils.IgnoreError(zipFile.Close)
 
-	// Create a temporary file and remove it, keeping the file descriptor.
-	tmpDir, err := os.MkdirTemp("", definitionsBaseDir)
+	// Create a temporary file and remove it for the OS to clean up once the
+	// struct is closed.
+	tmpFile, err := os.CreateTemp(h.rootDir, fileName)
```

Comment:
I thought the filename would be used by the HTTP handler when setting headers. That's why, originally, it was being created under a temporary directory with the same name as the ZIP's filename.

Reply:
Also, it should be unique per request. So, creating a temporary directory under the data directory (aka root dir) is not very beneficial; we might as well create it under `/tmp`. And then no need for making this a method of `h`.